### PR TITLE
feat(runner): fix '.only()' exclusive feature, #1481

### LINKED ISSUE #1481: describe.only matches more that one describe test suite
If I have these two test suites (just showing the first one)

```coffee
describe 'Boolean',->

  it 'is_True',->
    expect(true.is_True    ).to.be.an('Function')
    expect(false.is_True   ).to.be.an('Function')
    expect(true.is_True()  ).to.be.true
    expect(false.is_True() ).to.be.false
    expect((1==1).is_True()).to.be.true
    expect((1==2).is_True()).to.be.false
    expect((1!=1).is_True()).to.be.false
``` 

and

```coffee
describe 'Assert | Boolean',->
  it 'assert_Is_True', ->
    true.assert_Is_True.assert_Is_Function()
    true.assert_Is_True().assert_Is_True()
    (-> false.assert_Is_True()).assert_Throws()
```

If I add **.only** to the first one : 
```coffee
describe.only 'Boolean',->
```

both will now be executed:

image

image

and I would expect only the first one to execute

I assuming that mocha behind the scenes is not keeping a pointer to the actual tests to execute, it is just keeping the name/string, which is then searched on all tests/describes loaded (is that correct?)

### PR DESCRIPTION
**Pull request by a8m**
This PR fix #1481, and also extends the `.only()` behaviour.
e.g: 
* it's not use `grep` anymore
* support suite, test-case or both.
* add the ability to run multiple `.only()` (on many files)

There's a tests for [bdd](https://github.com/a8m/mocha/blob/master/test/acceptance/misc/only/bdd.js), [tdd](https://github.com/a8m/mocha/blob/master/test/acceptance/misc/only/tdd.js) and [qunit](https://github.com/a8m/mocha/blob/master/test/acceptance/misc/only/qunit.js) interfaces that illustrate the point more clearly. so please take a look on it, and let me now what do you think.

### CODE REVIEW

**Comment by boneskull** [comment]
@a8m Thanks!  Hopefully someone will check this out soon--I'm going to be swamped for next few wks still.  I'm thinking might be good for `3.0.0` release

**Comment by a8m** [comment]
Great @boneskull 
/cc @travisjeffery @dasilvacontin

**Comment by erezesk** [comment]
@a8m looks great, tyvm!
+1

**Comment by dasilvacontin** [comment]
Very busy lately, I'll try to dedicate some time to mocha this weekend. Anyways, thanks for the amazing work you've been pushing, @a8m! :)

**Comment by travisjeffery** [comment]
looks pretty good, i've looked it over quickly a couple of times and got the feeling some parts could be simpler/clearer. i should get sometime tomorrow to look more in-depth

**Comment by a8m** [comment]
Great @travisjeffery 
thx

**Comment by danielstjules** [comment]
Looks pretty good! Works with almost all my tests, though I think I found a couple edge cases. The first is with the root suite I think:

``` javascript
describe.only('suite', function() {
  it.only('test', function() {});

  it('test2', function() {});
});

it('should not be ran', function() {});

it.only('another test', function() {});
```

```
danielstjules:~/git/mocha (pr1591)
$ ./bin/mocha --reporter spec example.js


  ✓ should not be ran
  ✓ another test
  suite
    ✓ test


  3 passing (10ms)
```

On master, however, I get the following:

```
danielstjules:~/git/mocha (master =)
$ ./bin/mocha --reporter spec example.js


  ✓ another test

  1 passing (6ms)
```

I think it was decided in the issue that spec-related `only` just take precedence over suite level, so with that in mind, should the correct output of this suite be:

```
  ✓ another test
  suite
    ✓ test


  ` passing (11ms)
```

**Comment by danielstjules** [comment]
The second issue I hit was with the following test:

``` javascript
describe.only('suite', function() {
  it.only('test', function() {});

  it('test2', function() {});

  describe('suite2', function() {
    it('test3', function() {});
  });
});
```

```
danielstjules:~/git/mocha (pr1591)
$ ./bin/mocha --reporter spec example.js


  0 passing (3ms)
```

**Comment by a8m** [comment]
haha, I forgot to test the root suites, it should be fix [here](https://github.com/a8m/mocha/blob/master/lib/runner.js#L611). I'll do it.
about the second example,  I'll play with it and see what happens
thx @danielstjules

**Comment by a8m** [comment]
Fixed! 
PTAL @danielstjules

**Comment by danielstjules** [comment]
Those 2 look good, awesome!

Got another example, though I don't know if this is expected behavior or not. So this is probably not a bug, but just wanted to clarify :)

``` javascript
describe.only('suite', function() {
  describe('inner', function() {
    it.only('test', function() {});

    it('should not run', function() {});
  });

  it('should not run either', function() {});
});
```

```
dstjules:~/git/mocha (pr1591)
$ ./bin/mocha --reporter spec example.js


  suite
    ✓ should not run either
    inner
      ✓ test


  2 passing (8ms)
```

Is that expected output? If so, then the rest looks good to me!

**Comment by a8m** [comment]
This is the expected behaviour, PTAL [bdd-test](https://github.com/a8m/mocha/blob/master/test/acceptance/misc/only/bdd.js#L79)
Thanks for a great CR @danielstjules

**Comment by a8m** [comment]
This is the expected behaviour, PTAL [bdd-test](https://github.com/a8m/mocha/blob/master/test/acceptance/misc/only/bdd.js#L79)
Thanks for a agreat CR @danielstjules

**Comment by a8m** [comment]
rebased!

File: mocha.js [hunk 1]
```diff
@@ -5230,6 +5247,30 @@ Runner.prototype.abort = function(){
...
+ * @param {Array} suite
+ * @returns {Array}
+ * @api private
+ */
+function filterOnly(suite) {
+  // If it has `only` tests
+  if(suite.hasOnly) suite.tests = suite.hasOnly;
+  // If it doesn't has a suites make it running
+  // only if it mark as an `only`
+  if(!suite.suites.length) return suite.isOnly;
```
Comment (dasilvacontin) [inline mocha.js]: `suite.isOnly` is a boolean right? The function description say it returns `Array`.

File: mocha.js [hunk 2]
```diff
@@ -5230,6 +5247,30 @@ Runner.prototype.abort = function(){
...
+ * Filter suites based on `isOnly` logic.
+ *
+ * @param {Array} suite
+ * @returns {Array}
+ * @api private
+ */
+function filterOnly(suite) {
+  // If it has `only` tests
+  if(suite.hasOnly) suite.tests = suite.hasOnly;
+  // If it doesn't has a suites make it running
```
Comment (dasilvacontin) [inline mocha.js]: This comment has a couple of typos.

File: lib/runner.js [hunk 3]
```diff
@@ -657,6 +661,30 @@ Runner.prototype.abort = function(){
...
+ * Filter suites based on `isOnly` logic.
+ *
+ * @param {Array} suite
+ * @returns {Array}
+ * @api private
+ */
+function filterOnly(suite) {
+  // If it has `only` tests
+  if(suite.hasOnly) suite.tests = suite.hasOnly;
+  // If it doesn't has a suites make it running
```
Comment (dasilvacontin) [inline lib/runner.js]: This comment has a few typos.

File: lib/runner.js [hunk 4]
```diff
@@ -657,6 +661,30 @@ Runner.prototype.abort = function(){
...
+  // only if it mark as an `only`
+  if(!suite.suites.length) return suite.isOnly;
```
Comment (dasilvacontin) [inline lib/runner.js]: `suite.isOnly` is `Boolean` type, right? The function description says it returns `Array`.

It makes me wonder if it works correctly when this return is used, due to the following line:
```
if(this.hasOnly) rootSuite = filterOnly(rootSuite);
````

File: lib/runner.js [hunk 5]
```diff
@@ -657,6 +661,30 @@ Runner.prototype.abort = function(){
...
+  // Filter the nested suites
+  suite.suites = filter(suite.suites, filterOnly);
+  // `Unrunning` suites that not mark as an `only`
```
Comment (dasilvacontin) [inline lib/runner.js]: Maybe:
```javascript
//  Don't run suites that are not marked with `only`
```
Though if we are commenting the following line, maybe it would me more adequate:
```javascript
// Don't run tests from suites that are not marked as `only`
```

File: lib/runner.js (hunk 4 above)
Comment (a8m) [inline lib/runner.js]: It's used also as an iterator(recursively).

But, you right, it's a bug. I'll fix it quickly.
thx

File: lib/interfaces/tdd.js [hunk 6]
```diff
@@ -99,9 +101,12 @@ module.exports = function(suite){
...
 
     context.test.only = function(title, fn){
-      var test = context.test(title, fn);
-      var reString = '^' + escapeRe(test.fullTitle()) + '$';
-      mocha.grep(new RegExp(reString));
+      var test = context.test(title, fn)
+        , suite = test.parent;
+      mocha.options.hasOnly = true;
+      suite.isOnly = true;
+      suite.hasOnly = (suite.hasOnly || []).concat(test);
```
Comment (dasilvacontin) [inline lib/interfaces/tdd.js]: Isn't a bit confusing that you are using `hasOnly` as a `Boolean` in both `mocha.options` and `Runner`, and then you use it as an `Array` in `Suite`?

Besides that, the variable name `hasOnly` seems appropriate for a `Boolean`, but not so much for an `Array`.

e.g. `suite.tests` is the `Array` with the tests, `suite.onlyTests` could be the `Array` with the tests that 'are only'. It describes much better the content of the property.

File: lib/runner.js (hunk 4 above)
Comment (dasilvacontin) [inline lib/runner.js]: Okay, as expected, it does not work with the following test file (or is this not a valid test file?):
```javascript
it('should not be executed', function() { })
it.only('is the only test that should be executed', function() { })
it('shouldn\'t be executed either', function() { })
```

Comment (a8m) [inline lib/runner.js]: it's valid.
I'm on it.

File: lib/runner.js [hunk 7]
```diff
@@ -606,6 +607,9 @@ Runner.prototype.run = function(fn){
   var self = this,
     rootSuite = this.suite;
 
+  // If there is an `only` filter
+  if(this.hasOnly) rootSuite = filterOnly(rootSuite);
```
Comment (dasilvacontin) [inline lib/runner.js]: Okay, I see, you are also using`filterOnly` as a filter function. You only return `Boolean` or `suite`.
Why not change the line to:
```javascript
if(this.hasOnly) filterOnly(rootSuite);
```
You were always intending to return `rootSuite` here, right?

File: lib/runner.js [hunk 8]
```diff
@@ -657,6 +661,30 @@ Runner.prototype.abort = function(){
...
 
 /**
+ * Filter suites based on `isOnly` logic.
+ *
+ * @param {Array} suite
+ * @returns {Array}
+ * @api private
+ */
+function filterOnly(suite) {
+  // If it has `only` tests
```
Comment (dasilvacontin) [inline lib/runner.js]: ```// If it has `only` tests, we should only run those```

File: lib/runner.js (hunk 4 above)
Comment (dasilvacontin) [inline lib/runner.js]: Cool. I should reload more often, I wasn't seeing your comments. :)

**Comment by dasilvacontin** [comment]
@a8m, in general, you are writing `if(val)` instead of `if (val)`.

File: lib/runner.js [hunk 9]
```diff
@@ -657,6 +661,28 @@ Runner.prototype.abort = function(){
 };
 
 /**
+ * Filter suites based on `isOnly` logic.
+ *
+ * @param {Array} suite
+ * @returns {*}
```
Comment (dasilvacontin) [inline lib/runner.js]: Maybe we can set this now to `Boolean`?

**Comment by dasilvacontin** [comment]
LGTM :thumbsup:.

@travisjeffery, were you going to suggest any code improvements?

**Comment by a8m** [comment]
ping @mochajs/mocha

**Comment by dasilvacontin** [comment]
Well, goes.

**Comment by dasilvacontin** [comment]
We'll have to wait for the v3.0.0 release tho.

**Comment by boneskull** [comment]
@a8m I haven't gotten a chance to look at this yet, but please let me before this gets merged.

cc @mochajs/mocha

File: lib/interfaces/bdd.js [hunk 10]
```diff
@@ -95,9 +96,11 @@ module.exports = function(suite){
      */
 
     context.it.only = function(title, fn){
-      var test = context.it(title, fn);
-      var reString = '^' + escapeRe(test.fullTitle()) + '$';
-      mocha.grep(new RegExp(reString));
+      var test = context.it(title, fn)
```
Comment (boneskull) [inline lib/interfaces/bdd.js]: various places in this PR will need to be linted

**Comment by a8m** [comment]
Replaced with #1807(rebased on top of v3.0.0)

**Comment by Vanuan** [comment]
WTF? Are you trying to fix all bugs with one PR?

**Comment by boneskull** [comment]
@Vanuan I understand you are probably frustrated with the `.only()` bug(s), but please try to be constructive.  Thanks!
